weldx/asdf/tags/weldx/core/graph.py

```diff
@@ -0,0 +1,164 @@
+@dataclass
+class DiNode:
+    """Generic directed graph node type."""
+
+    edges: List["DiEdge"]
+    name: str = field(default_factory=uuid4)
+    attributes: dict = field(default_factory=dict)
+
+
+class DiNodeTypeASDF(WeldxType):
+    """ASDF type for `DiNode`."""
+
+    name = "core/graph/di_node"
+    version = "1.0.0"
+    types = [DiNode]
+    requires = ["weldx"]
+
+    @classmethod
+    def to_tree(cls, node: DiNode, ctx):
+        """Convert to python dict."""
+        if not node.attributes:
+            node.attributes = None
+        return node.__dict__
+
+    @classmethod
+    def from_tree(cls, tree, ctx):
+        """Reconstruct from tree."""
+        return DiNode(**tree)
+
+
+# Graph --------------------------------------------------------------------------------
+
+
+def build_tree(graph: nx.DiGraph, name: str, parent: str = None):
+    """Recursively build a tree structure of the graph starting from node ``name``.
+
+    Parameters
+    ----------
+    graph :
+        Complete graph to build.
+    name :
+        Name (or key) of the current node.
+    parent :
+        Key of the node
+
+    Returns
+    -------
+    DiNode
+        The root node object of the graph.
+
+    """
+    node = DiNode(name=name, edges=[])
+
+    for n in graph.neighbors(name):
+        if not n == parent:
+            child_node = build_tree(graph, n, parent=name)
+            if child_node:
+                edge = DiEdge(child_node, attributes=graph.edges[name, n])
+                node.edges.append(edge)
+    for n in graph.predecessors(name):
+        if not n == parent:
+            child_node = build_tree(graph, n, parent=name)
+            if child_node:
+                edge = DiEdge(
+                    child_node, attributes=graph.edges[n, name], direction="bwd"
+                )
+                node.edges.append(edge)
+    node.attributes = graph.nodes[name]  # add node attributes
+    return node
+
+
+def build_graph(current_node: DiNode, graph: nx.DiGraph = None):
+    """Recursively rebuild a (partial) graph from a DiNode object.
+
+    Parameters
+    ----------
+    current_node :
+        The current DiNode to be added to the graph.
+    graph :
+        The graph object that is being built.
+
+    Returns
+    -------
+    networkx.DiGraph
+        The constructed graph.
+
+    """
+    if graph is None:
+        graph = nx.DiGraph()
+
+    name = current_node.name
+    graph.add_node(name, **current_node.attributes)
+    for edge in current_node.edges:
+        attr = edge.attributes
+        if edge.direction == "bwd":
+            graph.add_edge(edge.target_node.name, name, **attr)
+        else:
+            graph.add_edge(name, edge.target_node.name, **attr)
+        build_graph(edge.target_node, graph)
+    return graph
+
+
+class DiGraphTypeASDF(WeldxType):
+    """Serialization class for `networkx.DiGraph`."""
+
+    name = "core/graph/di_graph"
+    version = "1.0.0"
+    types = [nx.DiGraph]
```

Comment:
So there is something worth noting here @marscher @vhirtham 

Since we have to register `nx.DiGraph` as a class for serialization to the weldx ASDF extension, `networkx` must be imported when `weldx` is loaded
currently I do not see a way to import this lazyly in this instance

Reply:
I see. I can revert the other lazy imports of networkx again, since they are then obsolete.

Reply:
we can do this at some point after this gets merged, no rush imo :)